File: python/koans/07_queue/test_dancers.py
import threading
import time

from koan_utils import EventLog, ThreadRunner, assert_blocks, eventually, jitter

from dancers import DanceFloor


def test_lone_leader_blocks():
    floor = DanceFloor()
    probe = assert_blocks(
        floor.leader_arrives, msg="a leader with no follower must wait"
    )
    threading.Thread(target=floor.follower_arrives, daemon=True).start()
    assert probe.wait(5), "the parked leader should proceed once a follower arrives"


def test_lone_follower_blocks():
    floor = DanceFloor()
    probe = assert_blocks(
        floor.follower_arrives, msg="a follower with no leader must wait"
    )
    threading.Thread(target=floor.leader_arrives, daemon=True).start()
    assert probe.wait(5), "the parked follower should proceed once a leader arrives"


def test_pair_completes():
    for _ in range(20):
        floor = DanceFloor()
        runner = ThreadRunner()
        runner.spawn(floor.leader_arrives, name="leader")
        runner.spawn(floor.follower_arrives, name="follower")
        runner.join_all(timeout=5)


def test_three_leaders_one_follower():
    floor = DanceFloor()
    done = EventLog()
    runner = ThreadRunner()

    def leader():
        floor.leader_arrives()
        done.record("leader")

    def follower():
        floor.follower_arrives()
        done.record("follower")

    for _ in range(3):
        runner.spawn(leader)
    time.sleep(0.2)
    runner.raise_worker_errors()
    assert done.count("leader") == 0, "no leader may proceed before any follower arrives"

    runner.spawn(follower)
    eventually(
        lambda: done.count("leader") == 1 and done.count("follower") == 1,
        timeout=5,
        msg="one follower should release exactly one pair",
    )
    time.sleep(0.3)  # a window for surplus leaders to sneak out
    assert done.count("leader") == 1, (
        f"one follower released {done.count('leader')} leaders"
    )

    runner.spawn(follower)
    runner.spawn(follower)
    runner.join_all(timeout=5)
    assert done.count("leader") == 3
    assert done.count("follower") == 3


def test_balanced_stress():
    floor = DanceFloor()
    done = EventLog()
    runner = ThreadRunner()

    def dancer(kind, arrive):
        jitter(3.0)
        arrive()
        done.record(kind)

    for _ in range(30):
        runner.spawn(dancer, "leader", floor.leader_arrives)
        runner.spawn(dancer, "follower", floor.follower_arrives)
    runner.join_all(timeout=15)
    assert done.count("leader") == 30
    assert done.count("follower") == 30

File: cpp/koans/07_queue/test_dancers.cpp
#include "koan_test.hpp"
#include "dancers.hpp"

#include <memory>
#include <thread>

using namespace koans;

KOAN_TEST(lone_leader_blocks) {
    auto floor = std::make_shared<DanceFloor>();
    auto probe = assert_blocks([floor] { floor->leader_arrives(); }, 300ms,
                               "leader_arrives (no follower yet)");
    std::thread([floor] { floor->follower_arrives(); }).detach();
    probe.assert_completed(5000ms, "the parked leader once a follower arrives");
}

KOAN_TEST(lone_follower_blocks) {
    auto floor = std::make_shared<DanceFloor>();
    auto probe = assert_blocks([floor] { floor->follower_arrives(); }, 300ms,
                               "follower_arrives (no leader yet)");
    std::thread([floor] { floor->leader_arrives(); }).detach();
    probe.assert_completed(5000ms, "the parked follower once a leader arrives");
}

KOAN_TEST(pair_completes) {
    for (int i = 0; i < 20; ++i) {
        auto floor = std::make_shared<DanceFloor>();
        ThreadRunner runner;
        runner.spawn([floor] { floor->leader_arrives(); }, "leader");
        runner.spawn([floor] { floor->follower_arrives(); }, "follower");
        runner.join_all(5000ms);
    }
}

KOAN_TEST(three_leaders_one_follower) {
    auto floor = std::make_shared<DanceFloor>();
    auto done = std::make_shared<EventLog>();
    ThreadRunner runner;
    auto leader = [floor, done] {
        floor->leader_arrives();
        done->record("leader");
    };
    auto follower = [floor, done] {
        floor->follower_arrives();
        done->record("follower");
    };
    for (int i = 0; i < 3; ++i) runner.spawn(leader, "leader");
    std::this_thread::sleep_for(200ms);
    KOAN_ASSERT_MSG(done->count("leader") == 0,
                    "no leader may proceed before any follower arrives");

    runner.spawn(follower, "follower");
    eventually(
        [&] { return done->count("leader") == 1 && done->count("follower") == 1; },
        5000ms, "one follower should release exactly one pair");
    std::this_thread::sleep_for(300ms);  // a window for surplus leaders
    KOAN_ASSERT_MSG(done->count("leader") == 1,
                    "one follower released " +
                        std::to_string(done->count("leader")) + " leaders");

    runner.spawn(follower, "follower");
    runner.spawn(follower, "follower");
    runner.join_all(5000ms);
    KOAN_ASSERT_EQ(done->count("leader"), static_cast<std::size_t>(3));
    KOAN_ASSERT_EQ(done->count("follower"), static_cast<std::size_t>(3));
}

KOAN_TEST(balanced_stress) {
    auto floor = std::make_shared<DanceFloor>();
    auto done = std::make_shared<EventLog>();
    ThreadRunner runner;
    for (int i = 0; i < 30; ++i) {
        runner.spawn(
            [floor, done] {
                jitter(3);
                floor->leader_arrives();
                done->record("leader");
            },
            "leader");
        runner.spawn(
            [floor, done] {
                jitter(3);
                floor->follower_arrives();
                done->record("follower");
            },
            "follower");
    }
    runner.join_all(15000ms);
    KOAN_ASSERT_EQ(done->count("leader"), static_cast<std::size_t>(30));
    KOAN_ASSERT_EQ(done->count("follower"), static_cast<std::size_t>(30));
}
